lib/updatehub/updatehub.c

```diff
@@ -0,0 +1,799 @@
+#else
+#define UPDATEHUB_SERVER "coap.updatehub.io"
+#endif
+
+static struct updatehub_context {
+	struct coap_block_context block;
+	struct k_sem semaphore;
+	struct flash_img_context flash_ctx;
+	struct tc_sha256_state_struct sha256sum;
+	enum updatehub_response code_status;
+	u8_t uri_path[MAX_PATH_SIZE];
+	u8_t payload[MAX_PAYLOAD_SIZE];
+	int downloaded_size;
+} _ctx;
+
+static struct update_info {
+	char package_uid[SHA256SUM_STRING_SIZE];
+	char sha256sum_image[SHA256SUM_STRING_SIZE];
+	int image_size;
+} _update_info;
+
+struct pollfd fds[1];
+static int sock;
+static int nfds;
+
+static void wait_fds(void)
+{
+	if (poll(fds, nfds, NETWORK_TIMEOUT) < 0) {
+		LOG_ERR("Error in poll");
+	}
+}
+
+static void prepare_fds(void)
+{
+	fds[nfds].fd = sock;
+	fds[nfds].events = 1;
+	nfds++;
+}
+
+static int metadata_hash_get(char *metadata)
+{
+	struct tc_sha256_state_struct sha256sum;
+	unsigned char hash[TC_SHA256_DIGEST_SIZE];
+	char buffer[2];
+
+	if (tc_sha256_init(&sha256sum) == 0) {
+		return -1;
+	}
+
+	if (tc_sha256_update(&sha256sum, metadata, strlen(metadata)) == 0) {
+		return -1;
+	}
+
+	if (tc_sha256_final(hash, &sha256sum) == 0) {
+		return -1;
+	}
+
+	memset(_update_info.package_uid, 0, SHA256SUM_STRING_SIZE);
+	for (int i = 0; i < TC_SHA256_DIGEST_SIZE; i++) {
+		snprintk(buffer, TC_SHA256_DIGEST_SIZE, "%02x",
+			 hash[i]);
+		strncat(&_update_info.package_uid[i], buffer, strlen(buffer));
+	}
+
+	return 0;
+}
+
+static bool
+is_compatible_hardware(struct resp_probe_some_boards *metadata_some_boards)
+{
+	int i;
+
+	for (i = 0; i < metadata_some_boards->supported_hardware_len; i++) {
+		if (strcmp(metadata_some_boards->supported_hardware[i],
+			   CONFIG_BOARD) == 0) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static void dns_cb(enum dns_resolve_status status, struct dns_addrinfo *info,
+		   void *user_ip)
+{
+	struct sockaddr_in *addr;
+	char hr_addr[NET_IPV6_ADDR_LEN], *hr_family, *buffer;
+
+	switch (status) {
+	case DNS_EAI_CANCELED:
+		LOG_ERR("DNS query was canceled");
+		goto error;
+	case DNS_EAI_FAIL:
+		LOG_ERR("DNS resolve failed");
+		goto error;
+	case DNS_EAI_NODATA:
+		LOG_ERR("Cannot resolve address");
+		goto error;
+	case DNS_EAI_ALLDONE:
+		break;
+	case DNS_EAI_INPROGRESS:
+		break;
+	default:
+		LOG_ERR("DNS resolving error");
+		goto error;
+	}
+
+	if (info && info->ai_family == AF_INET) {
+		hr_family = "IPv4";
+		addr = &net_sin(&info->ai_addr)->sin_addr;
+		buffer = net_addr_ntop(info->ai_family, addr, hr_addr, sizeof(hr_addr));
+		memcpy(user_ip, buffer, strlen(buffer));
+	}
+
+error:
+	k_sem_give(&_ctx.semaphore);
+}
+
+static bool start_coap_client(void)
+{
+	static struct sockaddr_in addr = { .sin_family = AF_INET,
+					   .sin_port = htons(5683) };
+	char ip[MAX_IP_SIZE];
+	u16_t dns_id;
+
+	/* Starts the semaphore to wait for the DNS callback */
+	k_sem_init(&_ctx.semaphore, 0, 1);
+
+	memset(&ip, 0, MAX_IP_SIZE);
+	if (dns_get_addr_info(UPDATEHUB_SERVER, DNS_QUERY_TYPE_A, &dns_id,
```

Comment:
BTW, the dns resolver function is async one. The function will return immediately, you need to add waiting of the results by yourself here.